File: AEDS-2/TP01/TP01Q17 - RECURSIVO - Palíndromo em C/palindromo.c
/*
   ==UserScript==
 @name         TP01Q17 - RECURSIVO - Palíndromo em C
 @namespace    https://github.com/Ddavidi/PUC-MINAS
 @description  VERDE PUC MINAS - TP01Q17 - RECURSIVO - Palíndromo em C
 @author       @ddavidi_
   ==/UserScript==
*/

#include <stdio.h>

int isEnd(char entrada[]){
    return entrada[0] == 'F' && entrada[1] == 'I' && entrada[2] == 'M' && entrada[3] == '\0';
}

int tamanhoEntrada(char entrada[]){
    int i = 0;

    while (entrada[i] != '\0'){
        i++;
    }

    return i;
}

int isPalindromo(char entrada[], int inicio, int fim){

    if(inicio >= fim){
        return 1;
    }

    if(entrada[inicio]!=entrada[fim]){
        return 0;
    }

    return isPalindromo(entrada, inicio+1, fim-1);
}

int isPalindromoAuxiliar(char entrada[]){
    int tamEntrada = tamanhoEntrada(entrada);

    if(tamEntrada == 0) return 1;

    return isPalindromo(entrada, 0, tamEntrada-1);
}

int main(void) {
    char entrada[500];

    scanf(" %[^\r\n]", entrada);

    while(!isEnd(entrada)){
        printf("%s\n", isPalindromoAuxiliar(entrada) ? "SIM" : "NAO");
        scanf(" %[^\r\n]", entrada);
    }

    return 0;
}

File: AEDS-2/TP01/TP01Q16 - RECURSIVO - Palíndromo/Palindromo.java
/*
   ==UserScript==
 @name         TP01Q16 - RECURSIVO - Palíndromo
 @namespace    https://github.com/Ddavidi/PUC-MINAS
 @description  VERDE PUC MINAS - TP01Q16 - RECURSIVO - Palíndromo
 @author       @ddavidi_
   ==/UserScript==
 */

import java.util.Scanner;

class Palindromo {
    public static boolean isPalindromo(String entrada, int inicio, int fim){
        if(inicio >= fim)
            return true;
        
        if(entrada.charAt(inicio)!=entrada.charAt(fim))
            return false;

        return isPalindromo(entrada, inicio+1, fim-1);
    }

    public static int tamanhoEntrada(String entrada){
        return entrada.length();
    }

    public static boolean isPalindromoAuxiliar(String entrada){
        int tamanhoEntrada = tamanhoEntrada(entrada);

        if(tamanhoEntrada == 0) return true;

        return isPalindromo(entrada, 0, tamanhoEntrada-1);
    }

    public static boolean isFim(String entrada){
        return entrada.charAt(0) == 'F' && entrada.charAt(1) == 'I' && entrada.charAt(2) == 'M';
    }

    public static void main(String[] args){
        Scanner scan = new Scanner(System.in);

        String entrada = scan.nextLine();

        while(!isFim(entrada)){
            System.out.println(isPalindromoAuxiliar(entrada) ? "SIM" : "NAO"); 
            entrada = scan.nextLine();
        }
    }
}

